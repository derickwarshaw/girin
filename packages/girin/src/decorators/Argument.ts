import { ArgumentMetadata } from '../metadata/ArgumentMetadata';
import { Generic } from '../metadata/Generic';


export interface ArgumentDecoratorOptions {
  defaultValue?: any;
  description?: string;
}

export function Argument(nameAndType: string, options?: ArgumentDecoratorOptions) {
  const [name, typeString] = nameAndType.split(/\s*:\s*/);

  return function(target: Object, fieldName: string, parameterIndex: number) {
    const definitionClass = target.constructor;
    ArgumentMetadata.create({
      definitionClass,
      fieldName,
      name,
      definedOrder: parameterIndex,
      generic: Generic.of(typeString),
      defaultValue: options && options.defaultValue,
      description: options && options.description,
    });
  }
}