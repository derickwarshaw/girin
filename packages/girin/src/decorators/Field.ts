import { FieldMetadata } from '../metadata/FieldMetadata';
import { Generic } from "../metadata/Generic";
import { globalMetadataStorage } from "../metadata/globalMetadataStorage";

export interface FieldDecoratorOptions {
  description?: string;
  deprecationReason?: string;
}

export function Field(type: string, options?: FieldDecoratorOptions): Function {
  return function(objectType: any, fieldName: string) {
    const definitionClass = objectType.constructor as any;
    FieldMetadata.create({
      fieldName,
      definitionClass,
      description: options && options.description,
      deprecationReason: options && options.deprecationReason,
      generic: Generic.of(type),
      meta: globalMetadataStorage,
    });
  };
}
