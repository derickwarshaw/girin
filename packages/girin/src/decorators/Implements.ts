import { MetadataStorage } from '../metadata/MetadataStorage';
import { ImplementsMetadata } from '../metadata/ImplementsMetadata';


export interface ImplementsDecoratorOptions {
  meta?: MetadataStorage;
}

export function Implements(targetNameOrDefinitionClass: Function | string, options: ImplementsDecoratorOptions = {}) {
  return function(definitionClass: Function & ImplementsDecoratorOptions) {
    const mergedConfig = Object.assign(options, definitionClass, { definitionClass, targetNameOrDefinitionClass });
    ImplementsMetadata.create(mergedConfig);
  };
}