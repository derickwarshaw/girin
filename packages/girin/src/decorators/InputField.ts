
import { InputFieldMetadata } from "../metadata/InputFieldMetadata";
import { MetadataStorage } from "../metadata/MetadataStorage";
import { Generic } from "../metadata/Generic";


export interface InputFieldDecoratorOptions {
  defaultValue?: any;
  description?: string;
  meta?: MetadataStorage;
}

export function InputField(nameAndType: string, options?: InputFieldDecoratorOptions): Function {
  const [ name, typeString ] = nameAndType.split(/\s*:\s*/);

  return function(definitionClass: Function, propertyName: undefined, parameterIndex: number) {
    InputFieldMetadata.create({
      definedOrder: parameterIndex,
      fieldName: name,
      generic: Generic.of(typeString),
      definitionClass: definitionClass,
      description: options && options.description,
      defaultValue: options && options.defaultValue,
      meta: options && options.meta,
    });
  };
}

