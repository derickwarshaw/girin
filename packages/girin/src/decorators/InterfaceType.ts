import { InterfaceTypeMetadata } from '../metadata/InterfaceTypeMetadata';
import { MetadataStorage } from '../metadata/MetadataStorage';
import { GraphQLFieldConfigMap, GraphQLInterfaceTypeConfig } from 'graphql';


export interface InterfaceDecoratorOptions {
  name?: string;
  meta?: MetadataStorage;
  fields?: () => GraphQLFieldConfigMap<any, any>;

  description?: GraphQLInterfaceTypeConfig<any, any>["description"];
  astNode?: GraphQLInterfaceTypeConfig<any, any>["astNode"];
  resolveType?: GraphQLInterfaceTypeConfig<any, any>["resolveType"];
}

export function InterfaceType(config: InterfaceDecoratorOptions = {}) {
  return function(definitionClass: InterfaceDecoratorOptions & Function) {
    const mergedConfig = Object.assign(config, definitionClass, { name: definitionClass.name, definitionClass });
    InterfaceTypeMetadata.create(mergedConfig);
  };
}
