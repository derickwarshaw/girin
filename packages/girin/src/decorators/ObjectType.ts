import { ObjectTypeMetadata } from '../metadata/ObjectTypeMetadata';
import { MetadataStorage } from '../metadata/MetadataStorage';
import { GraphQLObjectTypeConfig, GraphQLFieldConfigMap } from 'graphql';


export interface ObjectTypeDecoratorOptions {
  name?: string;
  meta?: MetadataStorage;
  fields?: () => GraphQLFieldConfigMap<any, any>;

  interfaces?: GraphQLObjectTypeConfig<any, any>["interfaces"];
  isTypeOf?: GraphQLObjectTypeConfig<any, any>["isTypeOf"];
  description?: GraphQLObjectTypeConfig<any, any>["description"];
  astNode?: GraphQLObjectTypeConfig<any, any>["astNode"];
  extensionASTNodes?: GraphQLObjectTypeConfig<any, any>["extensionASTNodes"];
}

export function ObjectType(options: ObjectTypeDecoratorOptions = {}) {
  return function(definitionClass: ObjectTypeDecoratorOptions & Function) {
    const mergedConfig = Object.assign(options, definitionClass, { name: definitionClass.name, definitionClass });
    ObjectTypeMetadata.create(mergedConfig);
  };
}
