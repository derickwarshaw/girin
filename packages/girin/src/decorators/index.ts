export { Argument } from './Argument';
export { Field } from './Field';
export { Implements } from './Implements';
export { InputField } from './InputField';
export { InputObjectType } from './InputObjectType';
export { ObjectType } from './ObjectType';
