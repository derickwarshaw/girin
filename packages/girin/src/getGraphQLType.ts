import { MetadataStorage } from "./metadata/MetadataStorage";
import { GraphQLType, isType } from 'graphql';

export interface TypeBuilderOptions {
  meta?: MetadataStorage;
}

export function getGraphQLType(definitionClassOrType: Function | GraphQLType, options?: TypeBuilderOptions): any {
  let meta: MetadataStorage;
  if (options && options.meta) {
    meta = options.meta;
  } else {
    meta = MetadataStorage.getMetadataStorage();
  }

  if (isType(definitionClassOrType)) {
    return definitionClassOrType;
  }

  const typeMetadata = meta.getDefinitionMetadata(definitionClassOrType);
  if (!typeMetadata) {
    throw new Error(`Given class ${definitionClassOrType.name} has no corresponding metadata`);
  }
  return typeMetadata.build.typeInstance;
}
