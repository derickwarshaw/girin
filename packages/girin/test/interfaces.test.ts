import { printSchema } from "graphql/utilities/schemaPrinter";
import { GraphQLSchema } from "graphql/type/schema";
import { graphql } from "graphql/graphql";

import { Field } from "../src/decorators/Field";
import { Implements } from "../src/decorators/Implements";
import { Argument } from "../src/decorators/Argument";
import { ObjectType } from "../src/decorators/ObjectType";
import { getGraphQLType } from "../src/getGraphQLType";
import { InterfaceType } from "../src/decorators/InterfaceType";


@InterfaceType()
class Person {
  @Field('String!') name: string;
  @Field('Int!') id: number;
}

const data: any = {
  celebrity: [
    { id: 0, name: 'Key', email: 'k@example.com' },
    { id: 1, name: 'Jonghyun', email: 'j@example.com' },
  ],
  fan: [
    { id: 0, name: 'Foo', email: 'foo@example.com', celebrityId: 0 },
    { id: 1, name: 'Bar', email: 'bar@example.com', celebrityId: 1 },
  ],
};

interface CelebritySource {
  id: number;
  name: string;
  email: string;
}

@ObjectType()
@Implements(Person)
class Celebrity extends Person {
  constructor(source: CelebritySource) {
    super();
    this.id = source.id;
    this.name = source.name;
    this.email = source.email;
  }



  @Field('String!') email: string;

  @Field('String!') name: string;
}

interface FanSource {
  id: number;
  name: string;
  email: string;
  celebrityId: number;
}

@ObjectType()
@Implements(Person)
class Fan extends Person {
  constructor(private source: FanSource) {
    super();
    this.id = source.id;
    this.email = source.email;
    this.name = source.name;
  }
  @Field('Int!') id: number;

  @Field('String!') email: string;

  @Field('String!') name: string;

  @Field('Celebrity')
  fanOf() {
    return new Celebrity(data.celebrity[this.source.celebrityId]);
  }
}

@ObjectType()
class Query {
  @Field('Person')
  person(
    @Argument('type: String!') type: string,
    @Argument('id: Int!') id: number,
  ) {
    const source = data[type].find((item: any) => item.id === id);
    if (type === 'celebrity') {
      return new Celebrity(source);
    } else if (type === 'fan') {
      return new Fan(source);
    }
    return null;
  }
}

const schema = new GraphQLSchema({
  query: getGraphQLType(Query),
  types: [
    getGraphQLType(Celebrity),
    getGraphQLType(Fan),
  ],
});


describe('interfaces', () => {
  it('generates schema as expected', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });

  it('resolves type', async () => {
    const result = await graphql({ schema, source: `
      query {
        celebrityQuery: person(type: "celebrity", id: 1) {
          id
          name
        }
        fanQuery: person(type: "fan", id: 0) {
          id
          name
          ...on Fan {
            fanOf {
              id
              name
            }
          }
        }
      }
    `});
    // console.log(JSON.stringify(result, null, '  '));

    expect(result.data!.celebrityQuery.id).toBe(1);
    expect(result.data!.celebrityQuery.name).toBe("Jonghyun");
    expect(result.data!.fanQuery.id).toBe(0);
    expect(result.data!.fanQuery.name).toBe("Foo");
    expect(result.data!.fanQuery.fanOf.id).toBe(0);
    expect(result.data!.fanQuery.fanOf.name).toBe("Key");
  });
});
