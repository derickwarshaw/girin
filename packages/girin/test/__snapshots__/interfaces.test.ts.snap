// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interfaces generates schema as expected 1`] = `
"type Celebrity implements Person {
  name: String!
  id: Int!
  email: String!
}

type Fan implements Person {
  name: String!
  id: Int!
  email: String!
  fanOf: Celebrity
}

interface Person {
  name: String!
  id: Int!
}

type Query {
  person(id: Int!, type: String!): Person
}
"
`;
